### PR TITLE
PICNIC-617: Show Username Following Colors In Menubar Widget

### PR DESCRIPTION
**Pull request by thebearjew**
Was broken for a while. Now uses `useRef` for a mutable box that won't trigger re-renders.
Purpose is to cache the list of username strings we've already sent to the remote window to prevent serializing and sending unnecessary data.

### Before
Screenshot 2019-11-06 10 45 07

### After

Screenshot 2019-11-06 10 45 24

### CODE REVIEW

File: shared/desktop/remote/sync-avatar-props.desktop.tsx [hunk 1]
```diff
@@ -99,12 +104,11 @@ function SyncAvatarProps(ComposedComponent: any) {
     windowParam: string
   }
 
-  class Wrapper extends React.PureComponent<WrapperProps, {usernames: Set<string>}> {
-    state = {usernames: new Set<string>()}
-    setUsernames = (usernames: Set<string>) => this.setState({usernames})
-    render() {
-      return <Connected {...this.props} usernames={this.state.usernames} setUsernames={this.setUsernames} />
```
Comment (thebearjew) [inline shared/desktop/remote/sync-avatar-props.desktop.tsx]: `setUsernames` was never being called. Looking back at the file history, I don't think using state makes sense anymore, so switching to `useRef` accomplished the same outcome.

`usernames` set is only used in `mapStateToProps` and that function should definitely not be updating state.

File: shared/common-adapters/usernames/remote-container.tsx [hunk 2]
```diff
@@ -6,12 +6,11 @@ import * as Container from './container'
 type OwnProps = Container.ConnectedProps
 
 // Connected username component
-const mapStateToProps = props => ({
-  _following: props.following,
-  _userInfo: props.userInfo,
-  _you: props.username,
+const mapStateToProps = state => ({
```
Comment (jakob223) [inline shared/common-adapters/usernames/remote-container.tsx]: lol

File: shared/desktop/remote/sync-avatar-props.desktop.tsx [hunk 3]
```diff
@@ -56,21 +54,28 @@ export const deserialize = (state: any = initialState, props: any) => {
...
-  const mapStateToProps = (state: Container.TypedState, ownProps: OwnProps) => ({
-    ...immutableCached(
-      getRemoteFollowers(state.config.followers, ownProps.usernames),
-      getRemoteFollowing(state.config.following, ownProps.usernames)
-    ),
-    httpSrvAddress: state.config.httpSrvAddress,
-    httpSrvToken: state.config.httpSrvToken,
-  })
+  const mapStateToProps = (state: Container.TypedState, ownProps: OwnProps) => {
+    const {
```
Comment (chrisnojima) [inline shared/desktop/remote/sync-avatar-props.desktop.tsx]: i find this nested destructuring hard to read. i think its clearer if you do it one level at a time aka
```
const {usernamesRef} = ownProps
const {current: oldUsername} = usernamesRef
```

Comment (jakob223) [reply shared/desktop/remote/sync-avatar-props.desktop.tsx]: This took me a long time to parse, I agree.

Comment (thebearjew) [reply shared/desktop/remote/sync-avatar-props.desktop.tsx]: Sure thing 👍

**Comment by thebearjew** [comment]
@chrisnojima Okay memoized the set creation and it's not thrashing.

Did some testing to make sure:

1. `getUsernameSet` is called once and only once
2. Debugging logs in the serializer show the `toSend` object only contains `following`/`followers` one and not on every `mapStateToProps` call. Tested this by keeping the widget open and spamming messages to a chat.

File: shared/desktop/remote/sync-avatar-props.desktop.tsx [hunk 4]
```diff
@@ -53,6 +53,15 @@ export const deserialize = (state: any = initialState, props: any) => {
   }
 }
 
+// Prevent <Connected /> from re-rendering on new Sets when usernames
+const getUsernamesSet = memoize(
+  (usernames: Array<string>) => {
+    console.log('JRY memoize usernamesSet')
```
Comment (chrisnojima) [inline shared/desktop/remote/sync-avatar-props.desktop.tsx]: 🔥

Comment (thebearjew) [reply shared/desktop/remote/sync-avatar-props.desktop.tsx]: everytime

**Pull request merged**